core/src/main/java/com/linecorp/armeria/internal/common/BuiltInDependencyInjector.java

```diff
@@ -15,40 +15,27 @@
  */
 package com.linecorp.armeria.internal.common;
 
-import static com.linecorp.armeria.internal.common.ReflectiveDependencyInjector.create;
-
 import java.util.Map;
-import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 
-import com.google.common.collect.ImmutableSet;
-
 import com.linecorp.armeria.common.DependencyInjector;
-import com.linecorp.armeria.server.annotation.ServerSentEventResponseConverterFunction;
-import com.linecorp.armeria.server.annotation.decorator.LoggingDecoratorFactoryFunction;
-import com.linecorp.armeria.server.annotation.decorator.RateLimitingDecoratorFactoryFunction;
 
 public enum BuiltInDependencyInjector implements DependencyInjector {
 
     INSTANCE;
 
-    //TODO(minwoox): Consider organizing built in class in a package and use reflection.
-    private static final Set<Class<?>> builtInClasses =
-            ImmutableSet.of(LoggingDecoratorFactoryFunction.class,
-                            RateLimitingDecoratorFactoryFunction.class,
-                            ServerSentEventResponseConverterFunction.class);
-
     private static final Map<Class<?>, Object> instances = new ConcurrentHashMap<>();
 
     @Override
     public <T> T getInstance(Class<T> type) {
-        if (!builtInClasses.contains(type)) {
+        //noinspection ConstantConditions
+        if (type.getDeclaredAnnotation(BuiltInDependency.class) == null) {
             return null;
         }
 
         //noinspection unchecked
         return (T) instances.computeIfAbsent(type, key -> {
-            final Object instance = create(key, null);
+            final Object instance = ReflectiveDependencyInjector.create(key, null);
```

Comment:
Logic looks good to me 👍  

What do you think of adding a test module under `:it` so that we don't forget to add this annotation?
We could import various modules and check if all injectable classes have the annotation:

```
final List<Class<?>> classes = ReflectionUtils.findAllClassesInPackage(
        "com.linecorp.armeria",
        ClassFilter.of(name -> !name.contains("Test"),
                       clazz -> {
                           return RequestConverterFunction.class.isAssignableFrom(clazz) &&
                                  Modifier.isPublic(clazz.getModifiers()) &&
                                  !Modifier.isAbstract(clazz.getModifiers()) &&
                                  !Modifier.isInterface(clazz.getModifiers());
                       }));
assertThat(classes).allMatch(clazz -> clazz.getDeclaredAnnotation(BuiltInDependency.class) != null);
```

I think it is also possible to expand this to other injectable classes like:
- ExceptionHandlerFunction
- RequestConverterFunction
- ResponseConverterFunction
- DecoratorFactoryFunction

Reply:
I think it's also possible to skip the test if it's too much work and just stick with adding the annotation to all affected classes for this iteration as well 🙏

Reply:
Added a test with a few modifications! 😆

Reply:
> ScalaPbRequestConverterFunction, ScalaPbResponseConverterFunction, FlowResponseConverterFunction) and didn't add @CreateIfMissing annotation on them.

The default instances of those converters are provided by SPI. Hence, we don't need to add `@CreateIfMissing` for them. By the way, it does not make any sense that some converters are created by SPI and others are created by reflections with `@CreateIfMissing`.

How about adding `{Request,Response}ConverterFunctionProvider` for all built-in converters? Users may choose either `@CreateIfMissing` or SPI style. Other maintainers might have a different opinion on this. Let's wait for their opinion before we proceed further.

Reply:
> By the way, it does not make any sense that some converters are created by SPI and others are created by reflections with @CreateIfMissing.

Agree with this point. As I've mentioned in [the first comment in the PR](https://github.com/line/armeria/pull/4534#issuecomment-1313136157), I am bit reluctant to add `@CreateIfMissing` annotation because the behavior that `ReflectiveDependencyInjecton` is disable when `SpringDependencyInjector` is used was actually intended, which I thought was a bug.

It seems the approach doesn't give much benefit compared to the added complexity.

Reply:
I think `{Request, Response}ConverterFunctionProvider` serves a slightly different needs from what we are handling in this PR but as I've stated in the comment above, I think we should be cautious of adding a new API to not confuse users.

Reply:
I checked the relation between `*ConverterFunctionProvider` and `@*Converter`.

- `*ConverterFunctionProvider` is used to dynamically resolve a `*ConverterFunction` depending on the given type. It would be tricky to use DI in that case.
-  `@*Converter` is statically created when it is attached to a service.

SPI style is necessary to globally apply a `*ConverterFunction` to all services without an explicit `@*Converter` annotation.

> I am bit reluctant to add @CreateIfMissing annotation because the behavior that ReflectiveDependencyInjecton is disable when SpringDependencyInjector is used was actually intended, which I thought was a bug.

I agree that it is confusing that the code that was working well suddenly stops when Spring injection. I was a person who opposes using `ReflectiveDependencyInjecton` with Spring injection by default. Because most dependency injection frameworks raise exceptions if no bean is registered. If `ReflectiveDependencyInjecton` created objects implicitly, it would cause another confusion. Some people would report an issue like 

> `NoSuchBeanDefinitionException` was ignored and an unknown object was created instead.

Actually, I suggested `@CreateIfMission` idea. But, now, I can't admit it is a good idea. 😅 On second thought, I came up with an idea that specifies whether to use DI or reflection to add an option to the annotations listed in the Javadoc.
https://github.com/line/armeria/blob/91c787f0a087499453103480f0fc8ab01072f9ab/core/src/main/java/com/linecorp/armeria/common/DependencyInjector.java#L33-L35
```java
enum CreationMode {
   // Use Java reflection to create an object.
   REFLECTION,
   // DI should provide an explicit instance.
   INJECTION
}

// CreationMode.REFLECTION is set by default.
@ResponseConverter(FooResponseConverter.class)
// Use CreationMode.INJECTION to use an external DI.
@ResponseConverter(FooResponseConverter.class, CreationMode.INJECTION)
```

Reply:
Thanks for suggesting an alternative with rich detail!

But it still leaves me a few questions.

* Would introducing `CreationMode` imply that users should always specify `CreationMode.INJECTION` in the annotations to make their custom `DependencyInjector (e.g. SpringDependencyInjector)` work?
* If so, wouldn't this be a bit verbose just for the purpose of making custom dependency injectors work?
  The most concise java code would look like...:
  ```java
  @ResponseConverterFunction(value = FooResponseConverter.class, mode = CreationMode.Injection)
  public Foo foo() { ... }
  ```
* `armeria.enable-auto-injection` flag in the springboot integration won't just magically work without specifying a `CreationMode` in every annotation.

Here's another approach I think would be nice both for flexibility and clarify of the injection code:

* Use `ReflectiveDependencyInjector` as the default `DependencyInjector` (which is same as the current behavior).
* `ServerBuilder#dependencyInjector(DependencyInjector)` replaces the default `DependencyInjector` with the one passed as parameter, instead of chaining the one added later to the original one.
* Relocate `ReflectiveDependencyInjector` under a public package so that users can opt-in Armeria's default behavior even if they use their own custom `DependencyInjector`.

No confusion is introduced using a custom `DependencyInjector` with this approach and it's flexible as the users who want to preserve Armeria's default injection behavior can choose to opt-in the original behavior by chaining `ReflectiveDependencyInjector` to the custom one. We can completely remove built-in dependency concept, which can be source of confusion, and the annotation interface remain concise, too.

```java
// Everything is explicit here; which injectors will be used and their orders.
final DependencyInjector injector =
        new FooDependencyInjector().orElse(new ReflectiveDependencyInjector());

sb.dependencyInjector(injector);
```

Reply:
> Would introducing CreationMode imply that users should always specify CreationMode.INJECTION in the annotations to make their custom DependencyInjector (e.g. SpringDependencyInjector) work?

Users may optionally use `CreationMode.INJECTION` if an object needs to be injected from a dependency injection. Most request or response converters do not need external dependencies which come with the Spring bean lifecycle. 

In other words, `CreationMode.REFLECTION` can be used with a custom `DependencyInjector`. `ReflectiveDependencyInjector` is no longer the default `DependencyInjector`. I am leaning toward having no default `DependencyInjector`. If `CreationMode` is introduced, `ReflectiveDependencyInjector` just can be renamed to a `ReflectiveObjectFactory` and remove `DependencyInjector` interface from super.

> new FooDependencyInjector().orElse(new ReflectiveDependencyInjector());

It is a possible solution but I would hesitate to recommend this pattern to users because it can swallow exceptions raised former one such as `NoSuchBeanDefinitionException` and create an object using reflection. What if a user forgets to add the object as a bean by mistake?

Reply:
> BuiltInDependencyInjector is no longer needed with the suggested approach, too, right?

Right. Only no dependency injector is added by default.

> I would like to send another PR after closing this if you guys wouldn't mind.

For sure. You can freely close this and open a PR.

Reply:
@ikhoon Thanks a lot for the nice suggestion. 👍 
@ks-yim Really sorry for making you work twice. 🙇